pkg/controller/rds/dbinstance/setup.go

```diff
@@ -0,0 +1,373 @@
+	}
+	if pw == "" && cr.Spec.ForProvider.AutogeneratePassword {
+		pw, err = password.Generate()
+		if err != nil {
+			return errors.Wrap(err, "unable to generate a password")
+		}
+		if err := e.savePasswordSecret(ctx, cr, pw); err != nil {
+			return errors.Wrap(err, errSaveSecretFailed)
+		}
+	}
+	obj.MasterUserPassword = aws.String(pw)
+	obj.DBInstanceIdentifier = aws.String(meta.GetExternalName(cr))
+	if len(cr.Spec.ForProvider.VPCSecurityGroupIDs) > 0 {
+		obj.VpcSecurityGroupIds = make([]*string, len(cr.Spec.ForProvider.VPCSecurityGroupIDs))
+		for i, v := range cr.Spec.ForProvider.VPCSecurityGroupIDs {
+			obj.VpcSecurityGroupIds[i] = aws.String(v)
+		}
+	}
+	if len(cr.Spec.ForProvider.DBSecurityGroups) > 0 {
+		obj.DBSecurityGroups = make([]*string, len(cr.Spec.ForProvider.DBSecurityGroups))
+		for i, v := range cr.Spec.ForProvider.DBSecurityGroups {
+			obj.DBSecurityGroups[i] = aws.String(v)
+		}
+	}
+	return nil
+}
+
+func (e *custom) postCreate(ctx context.Context, cr *svcapitypes.DBInstance, out *svcsdk.CreateDBInstanceOutput, _ managed.ExternalCreation, err error) (managed.ExternalCreation, error) {
+	if err != nil {
+		return managed.ExternalCreation{}, err
+	}
+	conn := managed.ConnectionDetails{
+		xpv1.ResourceCredentialsSecretUserKey: []byte(aws.StringValue(cr.Spec.ForProvider.MasterUsername)),
+	}
+	pw, _, err := rds.GetPassword(ctx, e.kube, cr.Spec.ForProvider.MasterUserPasswordSecretRef, cr.Spec.WriteConnectionSecretToReference)
+	if err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, "cannot get password from the given secret")
+	}
+	if pw != "" {
+		conn[xpv1.ResourceCredentialsSecretPasswordKey] = []byte(pw)
+	}
+	if out.DBInstance.Endpoint != nil {
+		if aws.StringValue(out.DBInstance.Endpoint.Address) != "" {
+			conn[xpv1.ResourceCredentialsSecretEndpointKey] = []byte(*out.DBInstance.Endpoint.Address)
+		}
+		if aws.Int64Value(out.DBInstance.Endpoint.Port) > 0 {
+			conn[xpv1.ResourceCredentialsSecretPortKey] = []byte(strconv.FormatInt(*out.DBInstance.Endpoint.Port, 10))
+		}
+	}
+	return managed.ExternalCreation{
+		ConnectionDetails: conn,
+	}, nil
+}
+
+func (e *custom) preUpdate(ctx context.Context, cr *svcapitypes.DBInstance, obj *svcsdk.ModifyDBInstanceInput) error {
+	obj.DBInstanceIdentifier = aws.String(meta.GetExternalName(cr))
+	pw, pwchanged, err := rds.GetPassword(ctx, e.kube, cr.Spec.ForProvider.MasterUserPasswordSecretRef, cr.Spec.WriteConnectionSecretToReference)
+	if err != nil {
+		return err
+	}
+	if pwchanged {
+		obj.MasterUserPassword = aws.String(pw)
+	}
+	return nil
+}
+
+func preDelete(_ context.Context, cr *svcapitypes.DBInstance, obj *svcsdk.DeleteDBInstanceInput) (bool, error) {
+	obj.DBInstanceIdentifier = aws.String(meta.GetExternalName(cr))
+	obj.FinalDBSnapshotIdentifier = aws.String(cr.Spec.ForProvider.FinalDBSnapshotIdentifier)
+	obj.SkipFinalSnapshot = aws.Bool(cr.Spec.ForProvider.SkipFinalSnapshot)
+	return false, nil
```

Comment:
There are some workarounds/hacks in `RDSInstance` code that should be ported over so that we don't need to fix them again. For example, deletion needs to call an update for the reasons listed [here](https://github.com/crossplane/provider-aws/blob/5a380ad/pkg/controller/database/rdsinstance.go#L240) and should skip making the deletion call if it's already in `deleting` state.